index.go

```diff
@@ -585,3 +611,113 @@ type importData struct {
 	RowIDs    []uint64
 	ColumnIDs []uint64
 }
+
+// CreateInputDefinition creates a new input definition.
+func (i *Index) CreateInputDefinition(pb *internal.InputDefinition) (*InputDefinition, error) {
+	// Ensure input definition doesn't already exist.
+	if i.inputDefinitions[pb.Name] != nil {
+		return nil, ErrInputDefinitionExists
+	}
+	return i.createInputDefinition(pb)
+}
+
+func (i *Index) createInputDefinition(pb *internal.InputDefinition) (*InputDefinition, error) {
+	if pb.Name == "" {
+		return nil, errors.New("input-definition name required")
+	} else if len(pb.Frames) == 0 || len(pb.Fields) == 0 {
+		return nil, errors.New("frames and fields are required")
+	}
+
+	for _, fr := range pb.Frames {
+		opt := FrameOptions{
+			RowLabel:       fr.Meta.RowLabel,
+			InverseEnabled: fr.Meta.InverseEnabled,
+			CacheType:      fr.Meta.CacheType,
+			CacheSize:      fr.Meta.CacheSize,
+			TimeQuantum:    TimeQuantum(fr.Meta.TimeQuantum),
+		}
+		_, err := i.CreateFrame(fr.Name, opt)
+		if err == ErrFrameExists {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+	}
+
+	// Initialize input definition.
+	inputDef, err := i.newInputDefinition(i.InputDefPath(), pb.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	inputDef.LoadDefinition(pb)
+	if err = inputDef.saveMeta(); err != nil {
+		return nil, err
+	}
+	i.inputDefinitions[pb.Name] = inputDef
+	return inputDef, nil
+}
+
+func (i *Index) newInputDefinition(path, name string) (*InputDefinition, error) {
+	f, err := NewInputDefinition(path, i.name, name)
+	if err != nil {
+		return nil, err
+	}
+	f.broadcaster = i.broadcaster
+	return f, nil
+}
+
+// DeleteInputDefinition removes a input definition from the index.
+func (i *Index) DeleteInputDefinition(name string) error {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	// Ignore if input definition doesn't exist.
+	inputDef := i.inputDefinition(name)
+	if inputDef == nil {
+		return nil
+	}
+
+	// Delete input definition file.
+	if err := os.Remove(filepath.Join(i.InputDefPath(), name)); err != nil {
+		return err
+	}
+
+	// Remove reference.
+	delete(i.inputDefinitions, name)
+	return nil
+}
+
+// openInputDefinition opens and initializes the input definitions inside the index.
+func (i *Index) openInputDefinition() error {
+	inputDefs, err := os.Open(i.path)
+	if err != nil {
+		return err
+	}
+	defer inputDefs.Close()
+
+	fis, err := inputDefs.Readdir(0)
+	if err != nil {
+		return err
+	}
+	for _, fi := range fis {
+		if fi.Name() != InputDefinitionRepo {
+			continue
+		}
+		inputPath, err := os.Open(i.InputDefPath())
+		if err != nil {
+			return err
+		}
+		inputFiles, err := inputPath.Readdir(0)
+		for _, file := range inputFiles {
+			input, err := i.newInputDefinition(i.InputDefPath(), file.Name())
+			if err != nil {
+				return err
+			}
+			input.Open()
+			i.inputDefinitions[file.Name()] = input
```

Comment:
is there a way we can share some of the logic from `createInputDefinition` around frame creation such that on startup opening the inputDefinition will create any frames that don't yet exist?